### PR TITLE
Add support for partition-local stores

### PR DESCRIPTION
**Pull request by tobim**
###  :notebook_with_decorative_cover: Description

This refactoring changes the protocol for the output path.
The EXPORTER is not an archive client any more but only talks to the index.
###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [ ] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

By commit.

### CODE REVIEW

File: libvast/vast/system/archive.hpp [hunk 1]
```diff
@@ -30,8 +30,8 @@ struct archive_state {
   void send_report();
   void next_session();
   archive_actor::pointer self;
-  std::unique_ptr<vast::store> store;
-  std::unique_ptr<vast::store::lookup> session;
+  std::unique_ptr<vast::segment_store> store;
+  std::unique_ptr<vast::segment_store::lookup> session;
```
Comment (dominiklohmann) [inline libvast/vast/system/archive.hpp:34]: Pretty sure these do not need to be wrapped in `unique_ptr` anymore. Up to you if you want to spend time on that, though.

File: libvast/vast/system/actors.hpp [hunk 2]
```diff
@@ -83,6 +83,22 @@ using status_client_actor = typed_actor_fwd<
...
 
+using store_actor = typed_actor_fwd<
+  // Register an exporter actor.
+  // TODO: This is only for compatibility with the legacy archive and
+  // can be removed with that.
+  caf::reacts_to<atom::exporter, caf::actor>,
+  // Starts handling a query for the given ids.
+  caf::reacts_to<ids, archive_client_actor>,
+  // Erase the events with the given ids.
+  caf::replies_to<atom::erase, ids>::with<atom::done>>::unwrap;
```
Comment (dominiklohmann) [inline libvast/vast/system/actors.hpp:94]: Should this be a status client actor as well?

File: libvast/vast/fbs/partition.fbs [hunk 3]
```diff
@@ -87,23 +64,23 @@ table v1 {
   /// TODO: Use the layout type once available.
   combined_layout: [ubyte];
 
-  /// A map storing the mapping from type name -> ids.
+  /// A map storing the mapping from type name -> ids
   type_ids: [type_ids.v0];
 
   /// Various synopsis structures for partition-wide synopses of certain
-  /// columns (e.g., global min and max timestamp).
+  /// columns. (eg. global min and max timestamp)
```
Comment (dominiklohmann) [inline libvast/vast/fbs/partition.fbs]: Please undo the change.

File: libvast/vast/fbs/partition.fbs [hunk 4]
```diff
@@ -87,23 +64,23 @@ table v1 {
   /// TODO: Use the layout type once available.
   combined_layout: [ubyte];
 
-  /// A map storing the mapping from type name -> ids.
+  /// A map storing the mapping from type name -> ids
```
Comment (dominiklohmann) [inline libvast/vast/fbs/partition.fbs]: Please undo the change.

File: libvast/vast/fbs/partition.fbs [hunk 5]
```diff
@@ -87,23 +64,23 @@ table v1 {
...
-  /// columns (e.g., global min and max timestamp).
+  /// columns. (eg. global min and max timestamp)
   partition_synopsis: partition_synopsis.v0;
 
   /// The contained value indexes.
   indexes: [qualified_value_index.v0];
 
   /// A store identifier and header information.
-  store: store_header;
+  /// The
```
Comment (dominiklohmann) [inline libvast/vast/fbs/partition.fbs:78]: This line shouldn't be here.

File: libvast_test/vast/test/fixtures/filesystem.hpp [hunk 6]
```diff
@@ -26,7 +26,7 @@ struct filesystem {
       FAIL(vast::render(err));
   }
 
-  vast::path directory = "vast-unit-test";
+  const vast::path directory = "vast-unit-test";
```
Comment (dominiklohmann) [inline libvast_test/vast/test/fixtures/filesystem.hpp:29]: What motivated this change? I definitely agree with it, just curious what caused you to make it as part of this PR.

File: libvast/vast/system/evaluator.hpp [hunk 7]
```diff
@@ -58,6 +58,9 @@ struct evaluator_state {
   /// Points to the parent actor.
   evaluator_actor::pointer self;
 
+  /// Stores the actor for sendings results to.
```
Comment (dominiklohmann) [inline libvast/vast/system/evaluator.hpp:61]: Comment seems to be wrongly copied from the `partition_client_actor` directly below?

File: libvast/src/system/evaluator.cpp [hunk 8]
```diff
@@ -153,7 +153,17 @@ evaluator(evaluator_actor::stateful_pointer<evaluator_state> self,
   self->state.expr = std::move(expr);
   self->state.eval = std::move(eval);
   self->state.store = std::move(store);
+  self->send(self->state.store, atom::exporter_v,
+             caf::actor_cast<caf::actor>(self));
   return {
+    [self](table_slice slice) {
+      self->send(self->state.client, std::move(slice));
```
Comment (dominiklohmann) [inline libvast/src/system/evaluator.cpp:160]: Should this `return self->delegate` rather than `self->send`?

File: libvast/src/system/evaluator.cpp [hunk 9]
```diff
@@ -153,7 +153,17 @@ evaluator(evaluator_actor::stateful_pointer<evaluator_state> self,
...
+    },
+    [self](atom::done, caf::error) {
+      VAST_DEBUG("{} completed expression evaluation", self);
+      self->state.promise.deliver(atom::done_v);
```
Comment (dominiklohmann) [inline libvast/src/system/evaluator.cpp:164]: How we we handle errors here?

File: libvast/test/system/evaluator.cpp [hunk 10]
```diff
@@ -103,12 +103,22 @@ struct fixture : fixtures::deterministic_actor_system_and_events {
...
     ids result;
+    REQUIRE(!self->mailbox().empty());
+    self->receive([&](const ids& hits,
+                      vast::system::archive_client_actor&) { result |= hits; },
+                  [] {});
+    REQUIRE(self->mailbox().empty());
+    self->send(eval, atom::done_v, caf::error{});
+    run();
     bool got_done_atom = false;
+    size_t slices = 0;
```
Comment (dominiklohmann) [inline libvast/test/system/evaluator.cpp:119]: Please rename to `num_slices` to express the intent more clearly.

File: libvast/vast/segment_store.hpp [hunk 11]
```diff
@@ -16,6 +16,11 @@
 #include "vast/segment_builder.hpp"
 #include "vast/uuid.hpp"
 
+// std::vector<table_slice> needs the definition on older versions of libstdc++.
+#if VAST_GCC && __GNUC__ <= 8
+#  include "vast/table_slice.hpp"
```
Comment (dominiklohmann) [inline libvast/vast/segment_store.hpp:21]: Wondering how this ever worked in `fwd.hpp` then.

**Review by dominiklohmann — COMMENTED** [review]
This is great. I like the approach you're taking for this one, it's a good mix of flexible and minimally invasive. 🚀 

I think extending `vast.fbs.partition.v0` is the way to go, so these two commits can probably be squashed into one in the end. I also left some in-line comments.

File: libvast/vast/system/archive.hpp (hunk 1 above)
Comment (tobim) [reply libvast/vast/system/archive.hpp:34]: I'm aware, but I'm avoiding changes that aren't strictly necessary in order to keep the diff small at the moment.

Comment (dominiklohmann) [reply libvast/vast/system/archive.hpp:34]: Makes sense, I kinda figured that was the reasoning.

File: libvast/vast/system/actors.hpp (hunk 2 above)
Comment (tobim) [reply libvast/vast/system/actors.hpp:94]: Not until the old archive is deprecated. Otherwise we would get the same information as part of the status report of every legacy partition.

File: libvast/vast/fbs/partition.fbs (hunk 3 above)
Comment (dominiklohmann) [reply libvast/vast/fbs/partition.fbs]: Outdated, resolving.

File: libvast/vast/fbs/partition.fbs (hunk 4 above)
Comment (dominiklohmann) [reply libvast/vast/fbs/partition.fbs]: Outdated, resolving.

File: libvast/vast/segment_store.hpp (hunk 11 above)
Comment (tobim) [reply libvast/vast/segment_store.hpp:21]: It happens in the definition of this data member.
```c++
caf::expected<std::vector<table_slice>> buffer_{caf::no_error};
```
I actually find it very surprising that the other compilers manage without the include here.

File: libvast/src/system/evaluator.cpp (hunk 9 above)
Comment (tobim) [reply libvast/src/system/evaluator.cpp:164]: I don't think escalating makes sense. I'd say just log `VAST_ERROR` and move on.

Comment (dominiklohmann) [reply libvast/src/system/evaluator.cpp:164]: Okay, was mostly wondering because we could deliver the error to the promise instead of the done.

File: libvast/src/system/evaluator.cpp (hunk 8 above)
Comment (tobim) [reply libvast/src/system/evaluator.cpp:160]: I'm not sure how delegate fits here?

File: libvast_test/vast/test/fixtures/filesystem.hpp (hunk 6 above)
Comment (tobim) [reply libvast_test/vast/test/fixtures/filesystem.hpp:29]: Maybe I shouldn't have, but the initialization code in the fixture for the index unit test just threw me off with the modification to this variable.

Comment (dominiklohmann) [reply libvast_test/vast/test/fixtures/filesystem.hpp:29]: It's fine to have in here. Was just curious. :-)

File: libvast/src/system/evaluator.cpp (hunk 8 above)
Comment (dominiklohmann) [reply libvast/src/system/evaluator.cpp:160]: The handler of the archive/store returns a `caf::result<void>`. You don't do error handling here, but the actor that sends the slices to the evaluator might want to do that. By delegating you forward errors that you receive.

File: libvast/vast/system/actors.hpp [hunk 12]
```diff
@@ -441,6 +430,7 @@ CAF_END_TYPE_ID_BLOCK(vast_actors)
 #define vast_uuid_synopsis_map std::map<vast::uuid, vast::partition_synopsis>
 CAF_ALLOW_UNSAFE_MESSAGE_TYPE(std::shared_ptr<vast_uuid_synopsis_map>)
 CAF_ALLOW_UNSAFE_MESSAGE_TYPE(std::shared_ptr<vast::partition_synopsis>)
+CAF_ALLOW_UNSAFE_MESSAGE_TYPE(vast::system::archive_request)
```
Comment (tobim) [inline libvast/vast/system/actors.hpp:433]: Add a type id if possible.

File: libvast/src/system/evaluator.cpp [hunk 13]
```diff
@@ -134,9 +134,7 @@ void evaluator_state::decrement_pending() {
   if (--pending_responses == 0) {
     // Now we ask the store for the actual data.
     // TODO: handle count estimate requests.
-    self->send(store, atom::exporter_v, caf::actor_cast<caf::actor>(client));
-    promise.delegate(store, atom::extract_v, hits,
-                     static_cast<receiver<table_slice>>(client));
+    promise.deliver(hits);
```
Comment (tobim) [inline libvast/src/system/evaluator.cpp:137]: rename `decrement_pending` to something more fitting.

File: libvast/src/system/partition.cpp [hunk 14]
```diff
@@ -814,8 +824,18 @@ partition_actor::behavior_type passive_partition(
...
+      auto rp = self->make_response_promise<atom::done>();
+      self->send(self->state.store, atom::exporter_v,
+                 caf::actor_cast<caf::actor>(client));
+      self->request(eval, caf::infinite, atom::run_v)
+        .then(
+          [self, client, rp](const ids& hits) mutable {
+            rp.delegate(self->state.store, atom::extract_v, hits,
+                        static_cast<receiver<table_slice>>(client));
+          },
+          [rp](caf::error err) mutable { rp.deliver(std::move(err)); });
```
Comment (tobim) [inline libvast/src/system/partition.cpp:837]:
```
[rp](caf::error& err) mutable { rp.deliver(std::move(err)); });
```

File: libvast/vast/system/actors.hpp [hunk 15]
```diff
@@ -222,8 +224,8 @@ using index_actor = typed_actor_fwd<
   // Conform to the protocol of the STATUS CLIENT actor.
   ::extend_with<status_client_actor>::unwrap;
 
-using archive_request
-  = std::tuple<caf::typed_response_promise<atom::done>, receiver<table_slice>>;
+using archive_request = std::tuple<caf::typed_response_promise<atom::done>,
+                                   receiver<table_slice>, expression>;
```
Comment (tobim) [inline libvast/vast/system/actors.hpp:228]: Move the expression and the promise into the `unhandled_ids` map, and remove the early popping.

Comment (tobim) [reply libvast/vast/system/actors.hpp]: The archive state management has been simplified and this comment does not apply any more.

File: libvast/vast/table_slice.hpp [hunk 16]
```diff
@@ -243,6 +255,34 @@ class table_slice final {
   inline static std::atomic<size_t> num_instances_ = {};
 };
 
+// Attribute-specifier-seqs are not allowed in friend function declarations, so
```
Comment (tobim) [inline libvast/vast/table_slice.hpp:258]: Move this below `evaluate()`.

File: libvast/src/system/archive.cpp [hunk 17]
```diff
@@ -137,14 +137,12 @@ archive(archive_actor::stateful_pointer<archive_state> self,
...
-           receiver_actor<table_slice> requester,
-           bool preserve_ids) -> caf::result<atom::done> {
-      VAST_DEBUG("{} got extract with the query {} and {} hints [{},  {})",
-                 self, expr, rank(xs), select(xs, 1), select(xs, -1) + 1);
-      auto op = preserve_ids ? archive_state::operation::extract_with_ids
-                             : archive_state::operation::extract;
-      return file_request(self, op, std::move(expr), xs,
+    [self](vast::query query, const ids& xs,
+           receiver_actor<table_slice> requester) -> caf::result<atom::done> {
+      VAST_DEBUG("{} got extract request with the query {}"
```
Comment (tobim) [inline libvast/src/system/archive.cpp]: just `request`.

File: libvast/src/system/archive.cpp [hunk 18]
```diff
@@ -208,12 +212,15 @@ archive(archive_actor::stateful_pointer<archive_state> self,
             }
             break;
           }
-          case vast::query::verb::count:
           case vast::query::verb::count_estimate: {
+            // This should be handled by the partition instead.
+            VAST_ASSERT(false);
```
Comment (tobim) [inline libvast/src/system/archive.cpp:217]: Use `die("...")` instead of assertions.

File: libvast/vast/query_options.hpp [hunk 19]
```diff
@@ -18,27 +18,35 @@ namespace vast {
...
+constexpr query_options unified = historical | continuous;
 
 constexpr bool has_query_option(query_options haystack, query_options needle) {
   return (static_cast<uint32_t>(haystack) & static_cast<uint32_t>(needle)) != 0;
 }
 
 constexpr bool has_historical_option(query_options opts) {
-  return has_query_option(opts, historical);
+  return has_query_option(opts, historical)
+         || has_query_option(opts, historical_with_ids);
```
Comment (tobim) [inline libvast/vast/query_options.hpp:45]: Make sure that this function is used everywhere.

File: libvast/vast/query_options.hpp [hunk 20]
```diff
@@ -18,27 +18,35 @@ namespace vast {
 enum class query_options : uint32_t {
   none = 0x00,
   historical = 0x01,
-  continuous = 0x02
+  historical_with_ids = 0x02,
```
Comment (tobim) [inline libvast/vast/query_options.hpp:21]: Use a dedicated `with_ids` flag instead.

File: libvast/vast/system/query_processor.hpp [hunk 21]
```diff
@@ -129,11 +127,7 @@ class query_processor {
   // -- implementation hooks ---------------------------------------------------
 
   /// Processes incoming hits from the INDEX.
-  virtual void process_hits(const ids& hits);
-
-  /// Processes incoming done messages. The default implementation always
-  /// transitions to the idle state.
-  virtual void process_end_of_hits();
+  virtual void process_done();
```
Comment (tobim) [inline libvast/vast/system/query_processor.hpp:130]: Fix comment.

File: libvast/vast/system/query_processor.hpp [hunk 22]
```diff
@@ -94,9 +94,7 @@ class query_processor {
   void start(vast::query query, index_actor index);
 
   /// @pre `state() == collect_hits`
-  /// @pre `n > 0`
-  /// @pre `partitions_.received + n <= partitions_.total`
-  void request_more_hits(uint32_t n);
+  bool request_more_results();
```
Comment (tobim) [inline libvast/vast/system/query_processor.hpp:97]: Document the return value.

**Review by tobim — COMMENTED** [review]
reviewed with @dominiklohmann.

File: libvast/vast/query.hpp [hunk 23]
```diff
@@ -9,25 +9,84 @@
...
+    system::receiver_actor<table_slice> sink;
+    ids_policy policy = {};
+
+    friend bool operator==(const extract& lhs, const extract& rhs) {
+      return lhs.sink == rhs.sink && lhs.policy == rhs.policy;
+    }
+
+    template <class Inspector>
+    friend auto inspect(Inspector& f, extract& q) {
+      return f(caf::meta::type_name("query"), q.sink, q.policy);
```
Comment (dominiklohmann) [inline libvast/vast/query.hpp:46]: Wrong name

File: libvast/vast/query.hpp [hunk 24]
```diff
@@ -9,25 +9,84 @@
...
-  enum class verb : uint8_t { //
-    count,
-    count_estimate,
-    erase,
-    extract,
-    extract_with_ids
+  struct count {
+    enum mode { estimate, exact };
+    system::receiver_actor<uint64_t> sink;
+    mode mode_ = {};
```
Comment (dominiklohmann) [inline libvast/vast/query.hpp:23]: `enum mode mode`.

File: libvast/src/system/application.cpp [hunk 25]
```diff
@@ -108,6 +108,7 @@ auto make_export_command() {
     opts("?vast.export")
       .add<bool>("continuous,c", "marks a query as continuous")
       .add<bool>("unified,u", "marks a query as unified")
+      //.add<bool>("with-ids", "preserve event ids on export")
```
Comment (dominiklohmann) [inline libvast/src/system/application.cpp:111]: Remove

**Review by dominiklohmann — COMMENTED** [review]
Reviewed the last two commits, they mirror exactly what we talked about in the last review.

File: libvast/vast/system/archive.hpp [hunk 26]
```diff
@@ -29,18 +29,42 @@ namespace vast::system {
...
+    extract_with_ids
+  };
+
+  struct request_state {
+    request_state(receiver_actor<table_slice> sink_, operation op_,
+                  expression expr_,
+                  std::pair<ids, caf::typed_response_promise<atom::done>> ids_)
+      : sink{std::move(sink_)}, op{op_}, expr{std::move(expr_)} {
+      ids_queue.push(std::move(ids_));
+    }
```
[the remark this answers is not in the document]
Comment (tobim) [reply libvast/vast/system/archive.hpp]: `std::queue` does not have an `initializer_list` constructor, so this is necessary.
https://en.cppreference.com/w/cpp/container/queue/queue

File: libvast/src/system/evaluator.cpp (hunk 13 above)
Comment (tobim) [reply libvast/src/system/evaluator.cpp:137]: After thinking about it I am of the opinion that `decrement_pending` isn't such a bad name for this function.

File: libvast/vast/system/query_processor.hpp [hunk 27]
```diff
@@ -94,6 +94,7 @@ class query_processor {
   void start(vast::query query, index_actor index);
 
   /// @pre `state() == collect_hits`
+  /// @returns false if there are no more partition to schedule.
```
Comment (tobim) [inline libvast/vast/system/query_processor.hpp:97]: partitions

File: libvast/src/system/archive.cpp [hunk 28]
```diff
@@ -83,6 +48,67 @@ void archive_state::send_report() {
   self->send(accountant, std::move(r));
 }
 
+std::unique_ptr<segment_store::lookup> next_session(archive_state& state) {
```
Comment (dominiklohmann) [inline libvast/src/system/archive.cpp:51]: Please make this a member function of the archive state and document it accordingly.

File: libvast/src/system/archive.cpp [hunk 29]
```diff
@@ -83,6 +48,67 @@ void archive_state::send_report() {
...
+    state.active_promise = std::move(rp);
+    state.session_ids = std::move(ids);
+    request.ids_queue.pop();
+    return state.store->extract(state.session_ids);
+  }
+  VAST_TRACE("archive has no requests");
+  return nullptr;
+}
+
+auto file_request(archive_actor::stateful_pointer<archive_state> self,
```
Comment (dominiklohmann) [inline libvast/src/system/archive.cpp:74]: Please make this a member function of the archive state and document it accordingly.

File: libvast/src/system/evaluator.cpp (hunk 13 above)
Comment (dominiklohmann) [reply libvast/src/system/evaluator.cpp:137]: Alrighty. I have no hard feelings on this.

File: libvast/vast/query.hpp [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    template <class Inspector>
+    friend auto inspect(Inspector& f, count& x) {
+      return f(caf::meta::type_name("vast.query.count"), x.sink, x.mode);
+    }
+  };
+
+  struct extract {
+    enum ids_policy { drop, preserve };
```
Comment (dominiklohmann) [inline libvast/vast/query.hpp:36]: Does it make sense to name this one `mode` as well, renaming the values to `drop_ids` and `preserve_ids`?

File: libvast/vast/system/actors.hpp [hunk 31]
```diff
@@ -80,41 +80,34 @@ using flush_listener_actor = typed_actor_fwd<
...
-  caf::reacts_to<ids>>::unwrap;
-
-/// The INDEX CLIENT actor interface.
-using index_client_actor = typed_actor_fwd<
-  // Receives done from the INDEX when the query finished.
-  caf::reacts_to<atom::done>>
-  // Receives ids from the INDEX for partial query hits.
-  ::extend_with<partition_client_actor>::unwrap;
+template <class T, class... Ts>
+using receiver_actor = typename typed_actor_fwd<
```
Comment (dominiklohmann) [inline libvast/vast/system/actors.hpp:84]: Missing comment.

File: libvast/vast/system/actors.hpp [hunk 32]
```diff
@@ -80,41 +80,34 @@ using flush_listener_actor = typed_actor_fwd<
...
+  // Add a new source.
+  typename caf::reacts_to<T, Ts...>>::unwrap;
 
 /// The STATUS CLIENT actor interface.
 using status_client_actor = typed_actor_fwd<
   // Reply to a status request from the NODE.
   caf::replies_to<atom::status, status_verbosity>::with< //
     caf::dictionary<caf::config_value>>>::unwrap;
 
+using store_actor = typed_actor_fwd<
```
Comment (dominiklohmann) [inline libvast/vast/system/actors.hpp:94]: Missing comment.

File: libvast/vast/system/actors.hpp [hunk 33]
```diff
@@ -80,41 +80,34 @@ using flush_listener_actor = typed_actor_fwd<
...
+  // Handles an extraction for the given expression, optionally optimized by a
+  // set of ids to pre-select the events to evaluate.
+  caf::replies_to<query, ids>::with<atom::done>,
+  // Erase the events with the given ids.
+  caf::replies_to<atom::erase, ids>::with<atom::done>>::unwrap;
+
+using store_builder_actor = typed_actor_fwd<>::extend_with<store_actor>
```
Comment (dominiklohmann) [inline libvast/vast/system/actors.hpp:101]: Missing comment.

File: libvast/vast/table_slice.hpp [hunk 34]
```diff
@@ -211,6 +211,18 @@ class table_slice final {
   friend void select(std::vector<table_slice>& result, const table_slice& slice,
                      const ids& selection);
 
+  /// Produces a new table slice consisting only of events addressed in `hints`
+  /// that match the given expression. Does not preserve ids, use `select
+  ///`instead if the id mapping must be maintained.
```
Comment (dominiklohmann) [inline libvast/vast/table_slice.hpp:216]:
```suggestion
  /// that match the given expression. Does not preserve ids; use `select`
  /// instead if the id mapping must be maintained.
```

File: libvast_test/vast/test/fixtures/actor_system.hpp [hunk 35]
```diff
@@ -21,6 +21,8 @@
 #include <caf/scoped_actor.hpp>
 #include <caf/test/dsl.hpp>
 #include <caf/test/io_dsl.hpp>
+// Defined in dsl.hpp and clashes with range_map::inject
```
Comment (dominiklohmann) [inline libvast_test/vast/test/fixtures/actor_system.hpp:24]:
```suggestion

// Defined in dsl.hpp and clashes with range_map::inject
```

**Review by dominiklohmann — COMMENTED** [review]
Just some minor code nits. This is looking really good now.

Doing some final regression testing locally before approving.

File: libvast/vast/query.hpp (hunk 30 above)
Comment (tobim) [reply libvast/vast/query.hpp:36]: Makes sense.

**Review by dominiklohmann — COMMENTED** [review]
Aside from the `count` performance regression while imports are running this is looking good. Looks like it'll be exactly 42 commits after all.

**Review by dominiklohmann — APPROVED** [review]
Performance is looking much better now while imports are running. IMO, this is ready to ship. 🚀
